source/extensions/tracers/datadog/tracer.cc

```diff
@@ -0,0 +1,126 @@
+#include "source/extensions/tracers/datadog/tracer.h"
+
+#include <cassert>
+#include <memory>
+#include <utility>
+
+#include "envoy/tracing/trace_context.h"
+
+#include "source/common/config/utility.h"
+#include "source/common/tracing/null_span_impl.h"
+#include "source/extensions/tracers/datadog/agent_http_client.h"
+#include "source/extensions/tracers/datadog/dict_util.h"
+#include "source/extensions/tracers/datadog/event_scheduler.h"
+#include "source/extensions/tracers/datadog/logger.h"
+#include "source/extensions/tracers/datadog/span.h"
+#include "source/extensions/tracers/datadog/time_util.h"
+
+#include "datadog/dict_reader.h"
+#include "datadog/error.h"
+#include "datadog/sampling_priority.h"
+#include "datadog/span_config.h"
+#include "datadog/trace_segment.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace Datadog {
+namespace {
+
+ThreadLocal::ThreadLocalObjectSharedPtr makeThreadLocalTracer(
+    datadog::tracing::TracerConfig config, Upstream::ClusterManager& cluster_manager,
+    const std::string& collector_cluster, const std::string& collector_reference_host,
+    TracerStats& tracer_stats, Event::Dispatcher& dispatcher, spdlog::logger& logger) {
+  config.logger = std::make_shared<Logger>(logger);
+  config.agent.event_scheduler = std::make_shared<EventScheduler>(dispatcher);
+  config.agent.http_client = std::make_shared<AgentHTTPClient>(
+      cluster_manager, collector_cluster, collector_reference_host, tracer_stats);
+
+  auto maybe_config = datadog::tracing::finalize_config(config);
+  if (auto* error = maybe_config.if_error()) {
+    auto prefix = "Unable to configure Datadog tracer. Tracing is now disabled. Error: ";
```

Comment:
I've never seen this pattern for declaring string constants. Maybe you could explicitly use whatever type with_prefix expects to avoid potentially implicitly converting between them?

Not that it matters for performance obviously.